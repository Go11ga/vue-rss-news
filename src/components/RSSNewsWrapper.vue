<template>
    <div class="rss-newswrapper">
        
        <template v-if="newsToShow.length">
            
            <RSSNewsItem  
                v-for="(item, i) in newsToShow"
                :key="i" 
                :model="item"
            />

            <RSSPagination :currentPage="currentPage" :totalCount="totalCount" />

        </template>

        <div class="rss-newswrapper__notfound" v-else>
            Ничего не найдено
        </div>

    </div>
</template>

<script>
import RSSNewsItem from '@/components/RSSNewsItem'
import RSSPagination from '@/components/ui/RSSPagination'
import { mapGetters, mapMutations } from 'vuex'

export default {
    name: 'RSSNewsWrapper',

    components: {
        RSSNewsItem,
        RSSPagination
    },

    data: () => ({
        newsPerPage: 4
    }),

    computed: {
        ...mapGetters([
            'getNewsToRender'
        ]),

        totalCount() {
            return Math.ceil(this.getNewsToRender.length / this.newsPerPage)
        },
        currentPage() {
            return +this.$route.params.pageNumber
        },
        newsToShow() {
            return [...this.getNewsToRender].splice((this.currentPage - 1) * this.newsPerPage, this.newsPerPage)
        }
    },

    methods: {
        ...mapMutations([
            'setActiveChannel'
        ])
    },

    watch: {
        getNewsToRender(news) {
            if (news.length === 0) {
                this.$toast.info('Поиск не дал результатов', {
                    timeout: 2000
                })
            }
        }
    },

    created() {
        this.setActiveChannel(this.$route.name)
    }
}
</script>