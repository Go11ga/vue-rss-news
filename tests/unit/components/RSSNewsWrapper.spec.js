import { shallowMount, createLocalVue } from '@vue/test-utils'
import RSSNewsWrapper from '@/components/RSSNewsWrapper'
import Vuex from 'vuex'
import VueRouter from 'vue-router'

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.use(VueRouter)

describe('RSSNewsWrapper.vue', () => {
    let wrapper
    let getters
    let mutations
    let store
    let router

    beforeEach(() => {
        getters = {
            getNewsToRender: () => [{}, {}]
        }

        mutations = {
            setActiveChannel: jest.fn()
        }

        store = new Vuex.Store({
            getters,
            mutations
        })

        const routes = [
            {
                path: '/mos/1',
                name: 'mos'
            }
        ]

        router = new VueRouter({
            routes
        })

        wrapper = shallowMount(RSSNewsWrapper, {
            localVue,
            store,
            router
        })
    })

    afterEach(() => {
        wrapper.destroy()
    })

    it('правильно рендерится', () => {
        expect(wrapper.vm.$options.name).toMatch('RSSNewsWrapper')

        expect(wrapper.html().includes('rss-newswrapper__notfound')).toBe(false)
    })

    it('snapshot', () => {
        expect(wrapper.element).toMatchSnapshot()
    })

    it('геттер totalCount', () => {
        expect(getters.getNewsToRender().length).toBe(2)
    })

    it('геттер currentPage', () => {
        let localThis = {
            $route: {
                params : {
                    pageNumber: 1
                }
            }
        }

        expect(RSSNewsWrapper.computed.currentPage.call(localThis)).toBe(1)
    })

    it('геттер newsToShow', () => {
        let localThis = {
            getNewsToRender: [
                { id: 1 },
                { id: 2 },
                { id: 3 },
                { id: 4 },
                { id: 5 },
                { id: 6 },
                { id: 7 },
                { id: 8 }
            ],
            currentPage: 1,
            newsPerPage: 4
        }

        expect(RSSNewsWrapper.computed.newsToShow.call(localThis)).toStrictEqual([ { id: 1 }, { id: 2 }, { id: 3 }, { id: 4 } ])
    })

    it('хук created', () => {
        expect(typeof RSSNewsWrapper.created).toBe('function')

        expect(mutations.setActiveChannel).toHaveBeenCalled()
        expect(mutations.setActiveChannel).toHaveBeenCalledTimes(1)
    })
})

describe('RSSNewsWrapper.vue', () => {
    let wrapper
    let getters
    let mutations
    let store
    let router

    beforeEach(() => {
        getters = {
            getNewsToRender: () => []
        }

        mutations = {
            setActiveChannel: jest.fn()
        }

        store = new Vuex.Store({
            getters,
            mutations
        })

        const routes = [
            {
                path: '/mos/1',
                name: 'mos'
            }
        ]

        router = new VueRouter({
            routes
        })

        wrapper = shallowMount(RSSNewsWrapper, {
            localVue,
            store,
            router
        })
    })

    afterEach(() => {
        wrapper.destroy()
    })

    it('правильно рендерится', () => {
        expect(wrapper.html().includes('rss-newswrapper__notfound')).toBe(true)
    })
})