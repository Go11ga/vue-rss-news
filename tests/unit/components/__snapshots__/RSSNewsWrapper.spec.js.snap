// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RSSNewsWrapper.vue snapshot 1`] = `
<div
  class="rss-newswrapper"
>
  <rssnewsitem-stub
    model="[object Object]"
  />
  <rssnewsitem-stub
    model="[object Object]"
  />
   
  <rsspagination-stub
    currentpage="NaN"
    totalcount="1"
  />
</div>
`;
